ducktape/tests/runner.py

```diff
@@ -14,216 +14,153 @@
-        """teardown method which stops services, gathers log data, removes persistent state, and releases cluster nodes.
-
-        Catch all exceptions so that every step in the teardown process is tried, but signal that the test runner
-        should stop if a keyboard interrupt is caught.
-        """
-        if teardown_services:
-            self.log(logging.INFO, "tearing down")
-
-        exceptions = []
-        if hasattr(self.current_test_context, 'services'):
-            services = self.current_test_context.services
-
-            # stop services
-            if teardown_services:
-                try:
-                    services.stop_all()
-                except BaseException as e:
-                    exceptions.append(e)
-                    self.log(logging.WARN, "Error stopping services: %s" % e.message + "\n" + traceback.format_exc(limit=16))
-
-            # always collect service logs
-            try:
-                self.current_test.copy_service_logs()
-            except BaseException as e:
-                exceptions.append(e)
-                self.log(logging.WARN, "Error copying service logs: %s" % e.message + "\n" + traceback.format_exc(limit=16))
-
-            # clean up stray processes and persistent state
-            if teardown_services:
-                try:
-                    services.clean_all()
-                except BaseException as e:
-                    exceptions.append(e)
-                    self.log(logging.WARN, "Error cleaning services: %s" % e.message + "\n" + traceback.format_exc(limit=16))
-
-        try:
-            self.current_test.free_nodes()
-        except BaseException as e:
-            exceptions.append(e)
-            self.log(logging.WARN, "Error freeing nodes: %s" % e.message + "\n" + traceback.format_exc(limit=16))
-
-        # Remove reference to services. This is important to prevent potential memory leaks if users write services
-        # which themselves have references to large memory-intensive objects
-        del self.current_test_context.services
-
-        if len([e for e in exceptions if isinstance(e, KeyboardInterrupt)]) > 0:
-            # Signal no more tests if we caught a keyboard interrupt
-            self.stop_testing = True
+        """Start a test runner client in a subprocess"""
+        proc = multiprocessing.Process(
+            target=run_client,
+            args=[
+                self.hostname,
+                self.port,
+                self.current_test_context.test_id,
+                self.current_test_context.logger_name,
+                self.current_test_context.results_dir,
+                self.session_context.debug,
+                self.session_context.max_parallel
+            ])
+        self.proc_list.append(proc)
+        proc.start()
+
+    def handle(self, event):
+        self.log(logging.DEBUG, str(event))
+
+        if event["event_type"] == ClientEventSupplier.READY:
```

Comment:
If the number of events stays small here, this approach works fine and is perfectly readable. If this grows substantially, you might just want to use a dictionary-based dispatch, i.e. you build a static dictionary like:

```
event_handlers = { ClientEventSupplier.READY: self.handle_ready, .... }
```

And then dispatch is just:

```
self.event_handlers[event["event_type"]](event)
```